src/pubsub/ua_pubsub_manager.c

```diff
@@ -0,0 +1,170 @@
+ */
+UA_StatusCode
+UA_Server_addPubSubConnection(UA_Server *server, const UA_PubSubConnectionConfig *connectionConfig,
+                              UA_NodeId *connectionIdentifier){
+    //iterate over the available UA_PubSubTransportLayers
+    for(size_t i = 0; i < server->config.pubsubTransportLayersSize; i++) {
+        if(connectionConfig && UA_String_equal(&server->config.pubsubTransportLayers[i].transportProfileUri, &connectionConfig->transportProfileUri)){
+            //create new connection config
+            UA_PubSubConnectionConfig *tmpConnectionConfig = (UA_PubSubConnectionConfig *) UA_calloc(1, sizeof(UA_PubSubConnectionConfig));
+            if(!tmpConnectionConfig){
+                UA_LOG_ERROR(UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
+                             "PubSub Connection creation failed. Out of Memory.");
+                return UA_STATUSCODE_BADOUTOFMEMORY;
+            }
+            //deep copy the given connection config
+            if(UA_PubSubConnectionConfig_copy(connectionConfig, tmpConnectionConfig) != UA_STATUSCODE_GOOD){
+                UA_LOG_ERROR(UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
+                             "PubSub Connection creation failed. Config copy problem.");
+                return UA_STATUSCODE_BADOUTOFMEMORY;
+            }
+            //create new connection and add to UA_PubSubManager
+            UA_PubSubConnection *newConnectionsField = (UA_PubSubConnection *)
+                    UA_realloc(server->pubSubManager.connections,
+                               sizeof(UA_PubSubConnection) * (server->pubSubManager.connectionsSize + 1));
+            if(!newConnectionsField) {
+                UA_PubSubConnectionConfig_deleteMembers(tmpConnectionConfig);
+                UA_free(tmpConnectionConfig);
+                UA_LOG_ERROR(UA_Log_Stdout, UA_LOGCATEGORY_SERVER,
+                             "PubSub Connection creation failed. Out of Memory.");
+                return UA_STATUSCODE_BADOUTOFMEMORY;
+            }
+            server->pubSubManager.connections = newConnectionsField;
+            UA_PubSubConnection *newConnection = &server->pubSubManager.connections[server->pubSubManager.connectionsSize];
+            memset(newConnection, 0, sizeof(UA_PubSubConnection));
+            newConnection->config = tmpConnectionConfig;
+            newConnection->channel = server->config.pubsubTransportLayers[i].createPubSubChannel(newConnection->config);
+            if(!newConnection->channel){
+                UA_PubSubConnection_delete(newConnection);
+                if(server->pubSubManager.connectionsSize > 0){
+                    newConnectionsField = (UA_PubSubConnection *)
+                            UA_realloc(server->pubSubManager.connections,
+                                       sizeof(UA_PubSubConnection) * (server->pubSubManager.connectionsSize));
+                    if(!newConnectionsField) {
+                        return UA_STATUSCODE_BADINTERNALERROR;
```

Comment:
The memory allocated for tmpConnectionConfig  will be leaked here I believe?

Reply:
The configuration copy is asigned in line 50 `newConnection->config = tmpConnectionConfig; ` and (in case of error) freed correctly in line 56 `UA_PubSubConnection_delete(newConnection); `.